### ISSUE TITLE
Headers compatibility

### ISSUE THREAD

**Issue by TimothyGu** [bitinn/node-fetch#181]
Currently, the semantics of Headers class is incompatible with the [Fetch Standard](https://fetch.spec.whatwg.org/). Merely fixing those incompatibilities is trivial, but _whether_ we want to do so needs a discussion, since

1. We might break existing applications using node-fetch
2. We might break compatibility with GitHub's [fetch](https://github.com/github/fetch) browser polyfill
3. We might break compatibility with existing native browser implementations (e.g. Google Chrome and Firefox)

I've compiled a list of potential issues, its behaviors on a specific platform, and my recommendation.

## Compatibility chart

Status quo:

Issue | Spec | GitHub | Browsers
--------|:------:|:-----------:|:------:|:-----:
ctor array | :x: | :white_check_mark: | :x:
ctor non-enum prop |    | :x: | :white_check_mark:
ctor proto prop |    | :white_check_mark: | Firefox :white_check_mark: Chrome :x:
ctor array prop | :x: | :x: | :x: 
append/set array | :x: | :x: | :x: 
`forEach`/`getAll` | :exclamation: | :white_check_mark: | :white_check_mark: 
`get` | :x: | :white_check_mark: | :white_check_mark: 
iterable | :x:  | :x:  | :x:
illegal chars | :x: | :white_check_mark: | :x:

Using the proposed default behavior:

Issue | Spec | GitHub | Browsers | Behavior changed?
--------|:------:|:-----------:|:------:|:-----:|:---------:
ctor array | :white_check_mark: | :x: | :white_check_mark: | Yes
ctor non-enum prop |    | :x: | :white_check_mark: | No
ctor proto prop |    | :white_check_mark: | Firefox :white_check_mark: Chrome :x:| No
ctor array prop | :white_check_mark:  | :white_check_mark: | :white_check_mark: | Yes
append/set array | :white_check_mark: | :white_check_mark: | :white_check_mark: | Yes
`forEach`/`getAll` | :exclamation: | :white_check_mark: | :white_check_mark: | No
`get` | :x: |  :white_check_mark: | :white_check_mark: | No
iterable | :x: | :white_check_mark: | :white_check_mark: | Yes
illegal chars | :white_check_mark: | :x: | :white_check_mark: | Yes

Using the proposed `FOLLOW_SPEC` mode:

Issue | Spec | GitHub | Browsers | Behavior changed?
--------|:------:|:-----------:|:------:|:-----:|:---------:
`get` | :white_check_mark: | :x: | :x: | Yes
iterable | :white_check_mark: | :x: | :x: | Yes

## Constructor with array

```js
var h = new Headers([['My-Header', 'first'], ['My-Header2', 'second']]);
console.log(h.get('My-Header'));
console.log(h.get('My-Header2'));
```

**[Spec](https://fetch.spec.whatwg.org/#dom-headers):** For every elements within the array, use the first element as the name and the second as the value (`'first'`; `'second'`)

**Current:** No support at all (`null`; `null`)

**GitHub:** No support at all (`null`; `null`)

**Chrome, Firefox:** Follows spec (`'first'`; `'second'`)

**My opinion:** Follows spec (`'first'`; `'second'`)

## Constructor with object with non-enumerable properties

```js
var a = {};
Object.defineProperties(a, {
  one: {enumerable: true, value: 'one'},
  two: {enumerable: false, value: 'two'},
});
var h = new Headers(a);
console.log(h.get('one'));
console.log(h.get('two'));
```

**[Spec](https://fetch.spec.whatwg.org/#concept-headers-fill):** Ambiguous (heycam/webidl#180)

**Current:** Ignores non-enumerable properties (`'one'`; `null`)

**GitHub:** Treats non-enumerable properties the same way as regular properties (`'one'`; `'two'`)

**Chrome & Firefox:** Ignores non-enumerable properties (`'one'`; `null`)

**My opinion:** Maintain current behavior (`'one'`; `null`)

## Constructor with object properties inherited from the prototype chain

```js
const a = Object.create({
  three: 'three'
});
const h = new Headers(a);
console.log(h.get('three'));
```

**[Spec](https://fetch.spec.whatwg.org/#concept-headers-fill):** Ambiguous (heycam/webidl#180)

**Current:** Does not include prototype properties (`null`)

**GitHub:** Does not include prototype properties (`null`)

**Chrome:** Does include prototype properties (`'three'`)

**Firefox:** Ignores non-enumerable properties (`null`)

**My opinion:** Until heycam/webidl#180 is resolved, maintain current behavior (`null`)

## Constructor with object properties whose values are arrays

```js
var h = new Headers({ 'My-Header': [ 'first', 'second' ] });
console.log(h.getAll('My-Header'));
```

**[Spec](https://fetch.spec.whatwg.org/#dom-headers-get):** The array needs to first be converted to `ByteString` using the `.toString()` method of the array, which is equivalent to joining with comma (`[ 'first,second' ]`)

**Current:** The individual elements are treated as separate headers (`[ 'first', 'second' ]`)

**GitHub:** Follows spec (`[ 'first,second' ]`)

**Chrome, Firefox:** Follows spec (`[ 'first,second' ]`)

**My opinion:** Follows spec (`[ 'first,second' ]`)

## `append()` and `set()` with array

```js
var h = new Headers({ 'My-Header': 'zeroeth' });
h.append('My-Header', ['first', 'second']);
h.set('My-Header2', ['first', 'second']);
console.log(h.getAll('My-Header'));
console.log(h.getAll('My-Header2'));

// Note: `getAll` is removed in spec, but for the express purpose of
// showing this problem let's pretend it still exists.
```

**[Spec](https://fetch.spec.whatwg.org/#dom-headers-get):** The array needs to first be converted to `ByteString` using the `.toString()` method of the array, which is equivalent to joining with comma (`[ 'zeroeth', 'first,second' ]`; `[ 'first,second' ]`)

**Current:** The conversion to string never happens, causing some inconsistent behaviors (`[ 'zeroeth', [ 'first', 'second' ] ]`, `[ [ 'first', 'second' ] ]`)

**GitHub:** Follows spec (`[ 'zeroeth', 'first,second' ]`; `[ 'first,second' ]`)

**Chrome, Firefox:** Follows spec (`[ 'zeroeth', 'first,second' ]`; `[ 'first,second' ]`)

**My opinion:** Follows spec (`[ 'zeroeth', 'first,second' ]`; `[ 'first,second' ]`)

## `forEach()` and `getAll()`

**[Spec](https://fetch.spec.whatwg.org/):** The two methods are removed.

**Current:** Still exist

**GitHub:** Still exist

**Chrome, Firefox:** Still exist

**My opinion:** Keep them

## `get()` with multiple values for the same header

```js
var h = new Headers;
h.append('My-Header', 'first');
h.append('My-Header', 'second');
console.log(h.get('My-Header'));
```

**[Spec](https://fetch.spec.whatwg.org/#dom-headers-get):** Return the "combined value" of all headers of the specified name, which means all headers of the name are joined using a comma (`'first,second'`)

**Current:** Returns the first header only (`'first'`)

**GitHub:** Returns the first header only (`'first'`)

**Chrome, Firefox:** Returns the first header only (`'first'`)

**My opinion:** Maintain current behavior by default, but offer an option to follow spec (`'first'`)

## Iterable interface

```js
var h = new Headers;
h.append('z', 'first');
h.append('a', 'second');
h.append('a', 'third');
h.append('z', 'fourth');
console.log(Array.from(h));
```

**[Spec](https://fetch.spec.whatwg.org/#dom-headers-set):** The "return value of running [sort and combine](https://fetch.spec.whatwg.org/#concept-header-list-sort-and-combine) with the header list." (`[ [ 'a', 'second,third' ], [ 'z', 'first,fourth' ] ]`)

**Current:** Not yet implemented; #180 implements the GitHub behavior

**GitHub:** Order of addition of header name, then of header (`[ [ 'z', 'first'], [ 'z', 'fourth' ], [ 'a', 'second' ], [ 'a', 'third' ] ]`)

**Chrome, Firefox:** Order of addition of header (`[ [ 'z', 'first'], [ 'a', 'second' ], [ 'a', 'third' ], [ 'z', 'fourth' ] ]`)

**My opinion:** Use GitHub's implementation, but offer an option to follow spec (`[ [ 'z', 'first'], [ 'z', 'fourth' ], [ 'a', 'second' ], [ 'a', 'third' ] ]`)

## Illegal characters in header name

```js
new Headers({ '\n': 'hey' });
```

**[Spec](https://fetch.spec.whatwg.org/#concept-headers-append):** Throws a `TypeError`

**Current:** Allowed

**GitHub:** Allowed

**Chrome & Firefox:** Follows spec; throws a `TypeError`

**My opinion:**  Follow spec; throw a `TypeError`

## Option to follow spec

For the following two issues, I propose adding an option either enabling the spec or (by default) disabling the spec behavior.

* `get()` with multiple values for the same header
* Iterable interface

This option could be module-wide, so to change it would be

```js
require('node-fetch/lib/headers.js').FOLLOW_SPEC = true;
```

**Comment by bitinn** [comment]
I have read through it and it looks good to me.

In general I think browsers did a good job following spec and keeping compatibility with existing code in the wild.

> get() with multiple values for the same header
>
> Spec: Return the "combined value" of all headers of the specified name, which means all headers of the name are joined using a comma ('first,second')

Looks like the spec changed, I have no idea why they did that, maybe during the drop of `getAll()`.

> Iterable interface

Now that I think of it, maybe there wasn't a sort-and-combine requirement in spec when I was implementing them, so this might again be a breaking spec change.

> Option to follow spec

I think this is the right approach, since spec changes rapidly (and doesn't seem to care about breaking changes). We should warn users, using this flag may face more breakage.

**Comment by bitinn** [comment]
One drawback with *Option to follow spec* approach is maintenance cost, but I guess you know that already.

**Comment by TimothyGu** [comment]
> One drawback with *Option to follow spec* approach is maintenance cost, but I guess you know that already.

Indeed. In #185 I made the test suite run twice -- once with the option enabled and once disabled. That should be able to help matters.

**Comment by TimothyGu** [comment]
Fixed in #185.
